File: problems/py01-softmax-dim0/baseline.py
"""题目 py01 的性能基线:直接对跨步维调用 PyTorch 自带的 softmax。

注意这个基线的性质 —— 它**不是**一份"写得差的实现",而是**用对了 API、
但撞上了坏的内存布局**:

    x 是 (rows, cols) 按行连续的张量,内存里同一列的元素相隔 cols 个 float。
    沿 dim=0 归约意味着每个线程要跨着取数据,几乎每次访问都吃不满一个
    cache line。PyTorch 的 softmax 内核本身写得很好,但它改变不了
    "归约维不连续"这件事。

所以这道题的解法不是"写个更好的 softmax",而是**重新安排内存布局,
让 PyTorch 的快路径能用上**。这是 PyTorch 使用者与 PyTorch 熟练者之间
最典型的一道分水岭。
"""
import torch


def forward(ctx):
    # 直接指定 dim=0 —— 这是最自然的写法,也是性能陷阱所在:
    # 对按行连续的张量来说,沿 dim=0 归约意味着跨步访存。
    return torch.softmax(ctx.x, dim=0)

File: problems/cpp01-matmul-loop-order/baseline.cpp
// 性能基线 —— 也是学习者拿到的起点代码。
//
// 这是**最自然的写法**:照着 C[i][j] = sum_k A[i][k]*B[k][j] 的定义直译。
// 它完全正确,数值上也挑不出毛病 —— 问题只出在访存顺序上。
//
// 具体地说,内层循环 k 递增时:
//   A[i*n+k] —— 连续访问,一个 cache line 装 16 个 float,摊下来每元素 0.25 次访存
//   B[k*n+j] —— **跨步访问**,步长 n*4 = 1536 字节,每读一个元素就换一条 cache line
//
// 于是内层循环每算一个乘加,就要从内存里搬一条 64 字节的 cache line,
// 而其中只有 4 个字节被用上 —— 有效利用率 1/16。

#include "ctx.h"

void matmul(Ctx& ctx) {
    const int n = ctx.n;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            float s = 0.0f;
            for (int k = 0; k < n; ++k) {
                s += ctx.A[i * n + k] * ctx.B[k * n + j];
            }
            ctx.C[i * n + j] = s;
        }
    }
}
